SilKit/source/services/metrics/MetricsTimerThread.cpp

```diff
@@ -0,0 +1,78 @@
+// SPDX-FileCopyrightText: 2024 Vector Informatik GmbH
+//
+// SPDX-License-Identifier: MIT
+
+#include "MetricsTimerThread.hpp"
+
+namespace VSilKit {
+
+MetricsTimerThread::MetricsTimerThread(std::function<void()> callback)
+    : _callback{std::move(callback)}
+    , _thread{MakeThread()}
+{
+}
+
+MetricsTimerThread::~MetricsTimerThread()
+{
+    if (!_started.exchange(true))
+    {
+        try
+        {
+            _go.set_value();
+        }
+        catch (const std::future_error &)
+        {
+        }
+    }
+
+    try
+    {
+        _done.set_value();
+    }
+    catch (const std::future_error &)
+    {
+    }
+
+    if (_thread.joinable())
+    {
+        _thread.join();
+    }
+}
+
+void MetricsTimerThread::Start()
+{
+    if (!_started.exchange(true))
+    {
+        try
+        {
+            _go.set_value();
+        }
+        catch (const std::future_error &)
+        {
+        }
+    }
+}
+
+auto MetricsTimerThread::MakeThread() -> std::thread
+{
+    auto go = _go.get_future();
+    auto done = _done.get_future();
+    return std::thread{[go = std::move(go), done = std::move(done), callback = &_callback]() mutable {
+        go.get();
+
+        while (true)
+        {
+            if (done.wait_for(std::chrono::seconds{1}) != std::future_status::timeout)
```

Comment:
If I got it right, the 1s here will be the update time for writing / sending the metrics. Don't we want to configure this?

Reply:
Yes that's correct. In the long run this will be a configuration setting, but for the initial experiment it doesn't have to be.